### PR TITLE
docs: explain new physical plan construction

### PR DESCRIPTION
**Pull request by crepererum**
There is no code yet, but this is the rough plan.

Ref #6098.

### CODE REVIEW

File: docs/physical_plan_construction.md [hunk 1]
```diff
@@ -0,0 +1,458 @@
...
+        predicate: (tag1 > 0) AND (some_fun(tag2) = 2)
+        files: ...
+```
+
+### Predicates & Record Batches
+`RecordBatchesExec` does not have any filter mechanism built in and hence relies on `FilterExec` to evaluate predicates.
+We therefore do NOT push down any predicates into `RecordBatchesExec`.
+
+## Projection Pushdown
+This concerns the pushdown of columns selections only. Note that `ProjectionExec` may contain renaming columns or even
```
Comment (alamb) [inline docs/physical_plan_construction.md:344]: similarly there is already a projection pushdown optimization that happens in DataFusion, but it is done for the LogicalPlan. I see that this is needed for the physical plans -- though given it is general it may also be something to consider contributing back up stream eventually

File: docs/physical_plan_construction.md [hunk 2]
```diff
@@ -0,0 +1,458 @@
...
+    child:
+      SomeExec1
+  - ProjectionExec:
+    keep: tag1, field2, time
+    child:
+      SomeExec2
+```
+
+### Projections & Predicates
+Projections may be pushed through `FilterExec` if the keep all columns required to evaluate the filter expression:
```
Comment (alamb) [inline docs/physical_plan_construction.md:373]:
```suggestion
Projections may be pushed through `FilterExec` if they keep all columns required to evaluate the filter expression:
```

**Review by alamb — APPROVED** [review]
TLDR is I think this is a great design @crepererum  -- well done 👏 

Did you consider moving more of the plan construction logic into LogicalPlans (rather than all in ExecutionPlans)? One benefit of doing that might be that the existing rules for projection pushdown and filter pushdown could possibly be reused. 

However, there are likely downsides to LogicalPlan construction as well (like the sorts might not be correctly added)

The only other final thing I would suggest thinking about is how much we can upstream to DataFusion (maybe some already exist in the physical optimizer https://github.com/apache/arrow-datafusion/tree/master/datafusion/core/src/physical_optimizer)

Another possible thought would be to possibly create

**Comment by crepererum** [comment]
> Did you consider moving more of the plan construction logic into LogicalPlans (rather than all in ExecutionPlans)?

Yes I did. I'll add a paragraph to the document, but I don't think it would work nicely. Our abstraction levels are:

- top-level we present `SchemaProvider` to DataFusion (implemented in `querier`, but not even used by the ingester/compactor, they use the `TableProvider` interface directly)
- `async SchemaProvider::table() -> TableProvider` is the last logical plan step (because then follows `async TableProvider::scan() -> ExecutionPlan`

Our `TableProvider` is (by DF) wrapped into a `DefaultTableSource` which ends up in a `LogicalPlan::TableScan`. So `Table{Provider,Source}::get_logical_plan` in theory allows a table to construct a "logical sub-plan", however this is only used by the `InlineTableScan` optimizer` if and only if there are NO filters applied to the table, which is rarely the case. So if I understand this whole wrapping and layering correctly[^1], I don't think DF really wants us to construct custom logical plans w/ custom nodes at the moment.

[^1]: I have currently an editor w/ over 10 files open just to understand how we get there starting w/ a SQL statement. It's a bit of an adventure.

File: docs/physical_plan_construction.md [hunk 3]
```diff
@@ -0,0 +1,458 @@
...
+            store: A
+        # if there are multiple stores (unlikely)
+        - ParquetExec:
+            store: B
+```
+
+The created `ParquetExec` does NOT contain any predicates or projections. The files may be grouped into
+`target_partitions` partitions.
+
+## Union Handling
```
Comment (NGA-TRAN) [inline docs/physical_plan_construction.md:55]: Form here is a part of `IOx Optimizer Passes`?

File: docs/physical_plan_construction.md [hunk 4]
```diff
@@ -0,0 +1,458 @@
...
+  - DeduplicateExec:
+      UnionExec:
+        - ChunkExec:
+            partition: A
+        - ChunkExec:
+            partition: A
+  - DeduplicateExec:
+      UnionExec:
+        - ChunkExec:
+            partition: B
```
Comment (NGA-TRAN) [inline docs/physical_plan_construction.md:119]: I like this one as you remove the SortPreservingergeExec that is currently unnecessary put after UnionExec to do the wok of union

File: docs/physical_plan_construction.md [hunk 5]
```diff
@@ -0,0 +1,458 @@
...
+        ts_max: 9
+    - ChunkExec:
+        ts_min: 11
+        ts_max: 15
+    - ChunkExec:
+        ts_min: 16
+        ts_max: 17
+    - ChunkExec:
+        ts_min: 17
+        ts_max: 18
```
Comment (NGA-TRAN) [inline docs/physical_plan_construction.md:150]: How exactly the DeduplicateExec works here after the Union? Do we need Union here or something else? My question is how do we keep data from all input streams sorted for deduplication?

File: docs/physical_plan_construction.md [hunk 6]
```diff
@@ -0,0 +1,458 @@
...
+      chunks: [C4, C7]
+      store: A
+  - ParquetExec:
+      chunks: [C5]
+      store: B
+```
+
+### Sort Grouping
+Since DataFusion will insert the necessary `SortExec`s for us, it is important that we are able to tell it about already
+sorted data. This only concerns `ParquetExec`, since `RecordBatchesExec` are based on not-yet-sorted ingester data.
```
Comment (NGA-TRAN) [inline docs/physical_plan_construction.md:237]: Oh, I do not know we allow `RecordBatchesExec` not sorted. I do not see any SortExec in today log. Is `RecordBatchesExec` not really sorted?

Question: Do we want to support `RecordBatchesExec` not sorted? And pros and cons? I see them so small and it does not hurt if the ingester does the sort job.

**Review by NGA-TRAN — APPROVED** [review]
The docs looks great. The projection and predicate pushdown I think sounds right. You make the plan look simpler and going from simple initial to further optimization 👍 .
The only detail I do not get is how deduplication works. Will you reuse the current mechanism? It is not clear to me how you keep input data sorted. However, those are details that can be sorted out as you go.

File: docs/physical_plan_construction.md (hunk 6 above)
Comment (alamb) [reply docs/physical_plan_construction.md:237]: > Question: Do we want to support RecordBatchesExec not sorted? And pros and cons? I see them so small and it does not hurt if the ingester does the sort job.

I think the ingester is so important to the rest of the system it should not be doing any sorting other than for persisting. If it does sorting for query as well, we run the risk of gumming up ingest with query which I think would be a significant problem

File: docs/physical_plan_construction.md [hunk 7]
```diff
@@ -0,0 +1,458 @@
...
+2. **IOx Optimizer Passes:** Apply IOx optimizer passes (in the order in which they occur in this document).
+3. **DataFusion Optimizer Passes:** Apply DataFusion's optimizer passes. This will esp. add all required sorts:
+   - before `DeduplicateExec`
+   - if output sorting is required (e.g. for SQL queries a la `SELECT ... FROM ... ORDER BY ...`)
+   - if any group-by expression within the DataFusion plan can use sorting
+
+This document uses YAML to illustrate the plans or plan transformations (in which case two plans are shown). Only the
+relevant parameters are shown.
+
+We assume that `QueryChunk`s can be transformed into `RecordBatchesExec`/`ParquetExec` and that we can always recover
```
Comment (wolffcm) [inline docs/physical_plan_construction.md:30]: I think that `RecordBatchesExec` nodes represent data that is coming from the ingesters, that has not yet been persisted to object storage. And `ParquetExec` on the other hand is querying object storage. Do I have that right?

File: docs/physical_plan_construction.md (hunk 6 above)
Comment (NGA-TRAN) [reply docs/physical_plan_construction.md:237]: Ah: Now I remember, this is the data communication between Ingester and Querier. Sounds good

File: docs/physical_plan_construction.md (hunk 7 above)
Comment (NGA-TRAN) [reply docs/physical_plan_construction.md:30]: > I think that RecordBatchesExec nodes represent data that is coming from the ingesters, that has not yet been persisted to object storage. And ParquetExec on the other hand is querying object storage. Do I have that right?

Yes, that is what I understand

File: docs/physical_plan_construction.md [hunk 8]
```diff
@@ -0,0 +1,458 @@
...
+            ts_max: 15
+  - DeduplicateExec:
+      UnionExec:
+        - ChunkExec:
+            ts_min: 16
+            ts_max: 17
+        - ChunkExec:
+            ts_min: 17
+            ts_max: 18
+```
```
Comment (wolffcm) [inline docs/physical_plan_construction.md:181]: Are `TimeSplit` and `PartitionSplit` mutually exclusive? If so how does the planner decide which one to use?

Comment (wolffcm) [reply docs/physical_plan_construction.md:181]: Thinking it over, I guess they are not mutually exclusive, you can apply them both, and you specify above that they are applied in the order they appear here. 👍

File: docs/physical_plan_construction.md (hunk 1 above)
Comment (wolffcm) [reply docs/physical_plan_construction.md:344]: Apache Calcite has a relational node called `Calc` that combines filter and projection. If we tend to see chains of projections and filters, I wonder if a similar operation would be useful in DataFusion:

https://calcite.apache.org/javadocAggregate/org/apache/calcite/rel/logical/LogicalCalc.html

**Review by wolffcm — APPROVED** [review]
Thanks for this write-up! I've found it educational.

File: docs/physical_plan_construction.md (hunk 8 above)
Comment (crepererum) [reply docs/physical_plan_construction.md:181]: > Thinking it over, I guess they are not mutually exclusive, you can apply them both, and you specify above that they are applied in the order they appear here. +1

That's correct :slightly_smiling_face:

File: docs/physical_plan_construction.md [hunk 9]
```diff
@@ -0,0 +1,458 @@
...
+that DataFusion knows about it (`FileScanConfig::output_ordering`).
+
+This somewhat interferes with the `target_partitions` setting. We shall find a good balance between avoiding resorts and
+"too wide" fan-outs.
+
+## Predicate Pushdown
+We may push down filters closer to the source under certain circumstances.
+
+### Predicates & Unions
+`FilterExec` can always be pushed through `UnionExec`, since `FilterExec` only allows row-based operations:
```
[the remark this answers is not in the document]
Comment (crepererum) [reply docs/physical_plan_construction.md:249]: Since these rules should run BEFORE the DF builtins (see first section about that) there shouldn't be any sorts at this point.

File: docs/physical_plan_construction.md (hunk 7 above)
Comment (crepererum) [reply docs/physical_plan_construction.md:30]: This is correct.

File: docs/physical_plan_construction.md (hunk 3 above)
Comment (crepererum) [reply docs/physical_plan_construction.md:55]: yes.

File: docs/physical_plan_construction.md (hunk 5 above)
Comment (crepererum) [reply docs/physical_plan_construction.md:150]: Sorts are introduced by DF after the IOx passes run. `DeduplicateExec` tells DF (already) that it wants its inputs sorted. Note that "sorting" is a per-DF-partition attribute (i.e. every stream that the `UnionExec` produces can have a different sorting). If `ChunkExec` (parquet or record batch) doesn't produce the sorted data that `DeduplicateExec` wants, DF will automatically insert a `SortExec` in-between. I HOPE (but have to verify) that the `SortExec` will be inserted below the `UnionExec` so it doesn't sort parquet files that are already correctly sorted (if not, I'll fix that upstream).

Does this answer your question?

File: docs/physical_plan_construction.md [hunk 10]
```diff
@@ -0,0 +1,458 @@
+# IOx Physical Plan Construction
+This document describes how DataFusion physical plans should be constructed in IOx. As a reminder: Our logical plans
+contain a `ChunkTableProvider` (implements `TableProvider`) that contains a set of `QueryChunk`s. The main entry point
```
[the remark this answers is not in the document]
Comment (crepererum) [reply docs/physical_plan_construction.md]: done (via a slightly different link structure)

File: docs/physical_plan_construction.md (hunk 5 above)
Comment (NGA-TRAN) [reply docs/physical_plan_construction.md:157]: You answer the sort part 👍 
What if the data is already sorted? Will Union keep all the streams of data in the order for us to do the merge and remove duplicate, or Union seemly concat the input streams which won't produce the result we want?

Comment (crepererum) [reply docs/physical_plan_construction.md:157]: `UnionExec` only forwards streams/partitions (e.g. if you have:

```yaml
---
UnionExec:
  - SomeExec:
      n_partitions: 1
  - SomeExec:
      n_partitions: 2
```

then the `UnionExec` will have 3 output partitions. It does not touch the data in these streams at all.
